src/clientHandler.ts

```diff
@@ -12,298 +10,177 @@ import {
       initializationOptions: initializationOptions,
       initializationFailedHandler: (error) => {
         this.reporter.sendTelemetryException(error);
         return false;
       },
-      outputChannel: outputChannel,
+      outputChannel: this.outputChannel,
       revealOutputChannelOn: RevealOutputChannelOn.Never,
     };
 
-    const client = new LanguageClient(id, name, serverOptions, clientOptions);
+    const id = `terraform`;
+    const client = new LanguageClient(id, serverOptions, clientOptions);
 
+    const enableReferenceCountCodeLens: boolean = config('terraform').get('enableReferenceCountCodeLens');
     if (enableReferenceCountCodeLens) {
       client.registerFeature(new ShowReferencesFeature(client));
     }
 
     client.onDidChangeState((event) => {
+      console.log(`Client: ${State[event.oldState]} --> ${State[event.newState]}`);
       if (event.newState === State.Stopped) {
-        clients.delete(location);
         this.reporter.sendTelemetryEvent('stopClient');
       }
     });
 
     return { commandPrefix, client };
   }
 
-  public async stopClients(folders?: string[]): Promise<void[]> {
-    const promises: Promise<void>[] = [];
+  private getServerOptions() {
+    const cmd = this.lsPath.resolvedPathToBinary();
+    const serverArgs: string[] = config('terraform').get('languageServer.args');
+    const executable: Executable = {
+      command: cmd,
+      args: serverArgs,
+      options: {},
+    };
+    const serverOptions: ServerOptions = {
+      run: executable,
+      debug: executable,
+    };
+    return serverOptions;
+  }
 
-    if (this.supportsMultiFolders) {
-      promises.push(this.stopClient(MULTI_FOLDER_CLIENT));
-      return Promise.all(promises);
-    }
+  private getInitializationOptions(commandPrefix: string) {
+    const rootModulePaths: string[] = config('terraform-ls').get('rootModules');
+    const terraformExecPath: string = config('terraform-ls').get('terraformExecPath');
+    const terraformExecTimeout: string = config('terraform-ls').get('terraformExecTimeout');
+    const terraformLogFilePath: string = config('terraform-ls').get('terraformLogFilePath');
+    const excludeModulePaths: string[] = config('terraform-ls').get('excludeRootModules');
+    const ignoreDirectoryNames: string[] = config('terraform-ls').get('ignoreDirectoryNames');
 
-    if (!folders) {
-      folders = [];
-      for (const key of clients.keys()) {
-        folders.push(key);
-      }
+    if (rootModulePaths.length > 0 && excludeModulePaths.length > 0) {
+      throw new Error(
+        'Only one of rootModules and excludeRootModules can be set at the same time, please remove the conflicting config and reload',
+      );
     }
 
-    for (const folder of folders) {
-      promises.push(this.stopClient(folder));
+    const experimentalFeatures = config('terraform-ls').get('experimentalFeatures');
+    let initializationOptions = { commandPrefix, experimentalFeatures };
```

Comment:
This might be easier to read and results in a better type for `initializationOptions `.

```ts
const initializationOptions = {
  commandPrefix,
  experimentalFeatures,
  ...(terraformExecPath.length > 0 && { terraformExecPath }),
  ...(terraformExecTimeout.length > 0 && { terraformExecTimeout }),
  ...
};
```